### ISSUE TITLE
lektor.db.Record equality

### ISSUE THREAD

**Issue by relikd** [lektor/lektor#1101]
Should two pages be equal even though the alt is different? I just stumbled upon this because I wanted to store source objects in a set but the hash and equality says it is the same object.

https://github.com/lektor/lektor/blob/74bce9096116caffe27ab69327ed4ba274e66e1e/lektor/db.py#L483-L488

**Comment by dairiki** [comment]
Agree, that does not seem right.  We should *probably* fix that.

I do get just a little queasy thinking about what subtle breakage might occur if we do.

For grins, I tried changing `Record.__eq__` to check (*path*, *alt*) for the equality key, and it didn't seem to break any of our tests, for whatever that's worth.

A quick look, though, found set operations on pages being performed here:

https://github.com/lektor/lektor/blob/74bce9096116caffe27ab69327ed4ba274e66e1e/lektor/build_programs.py#L224-L225

Those set operations *could* be affected by a change to page equality semantics — though in this specific case I don't *think* it would make a difference.  I don't think our tests currently have good coverage for this sort of thing, however.
 
## \_\_hash\_\_ is broken, too

https://github.com/lektor/lektor/blob/74bce9096116caffe27ab69327ed4ba274e66e1e/lektor/db.py#L493-L494

Here, `__hash__` depends on `self.path`, while `__eq__` looks at `self["_path"]`.  Those are potentially different, e.g. for `Page`s that have a  `page_num` set.

https://github.com/lektor/lektor/blob/74bce9096116caffe27ab69327ed4ba274e66e1e/lektor/db.py#L557-L562

This means that currently, two `Page`s can have different hashes, while still comparing equal. This violates [Python's requirement](https://docs.python.org/3/reference/datamodel.html#object.__hash__) that hashes for objects that compare equal must be equal.

In other words, for `Page`s, `page_num` should also be included in the equality key.

**Comment by dairiki** [comment]
I just found [this](https://github.com/lektor/lektor/pull/1007/commits/a2dc8ba8438fa573ad5d795f478a658cb7ddaa7d#diff-80e08515cf5d832e9854a388856e05133279ba2d2335f314983596e65ebbcafcL495-L510) is my year-old draft PR #1007.  (So, as I said, I guess I agree with you.) :laughing: 

I suppose I should put some more work into that PR.
